packages/url_launcher/url_launcher_web/lib/src/link.dart

```diff
@@ -197,41 +343,99 @@ class LinkViewController extends PlatformViewController {
-    // After the DOM click event has been received, clean up the hit test state
-    // so we can start fresh on the next event.
-    unregisterHitTest();
+    handleGlobalClick(
+      event: event,
+      target: event.target as html.Element?,
+    );
+  }
+
+  /// Global click handler that triggers on the `capture` phase. We use `capture`
+  /// because some events may be consumed and prevent further propagation at the
+  /// target. This may lead to issues (see: https://github.com/flutter/flutter/issues/143164)
+  /// where a followLink was executed but the event never bubbles back up to the
+  /// window (e.g. when button semantics obscure the platform view). We make sure
+  /// to only trigger the link if a hit test was registered and remains valid at
+  /// the time the click handler executes.
+  @visibleForTesting
+  static void handleGlobalClick({
+    required html.MouseEvent event,
+    required html.Element? target,
+  }) {
+    // We only want to handle clicks that land on *our* links, whether that's a
+    // platform view link or a semantics link.
+    final int? viewIdFromTarget = _getViewIdFromLink(target) ??
+        _getViewIdFromSemanticLink(target);
+
+    if (viewIdFromTarget == null) {
+      // The click target was not one of our links, so we don't want to
+      // interfere with it.
+      //
+      // We also want to reset the signals in this case.
+      _triggerSignals.reset();
+      return;
+    }
+
+    _triggerSignals.registerDomEvent(
+      viewId: viewIdFromTarget,
+      mouseEvent: event,
+    );
+
+    if (!_triggerSignals.isValid) {
+      // The view ID from the target doesn't match the view ID from the follow
+      // link. Let's prevent the browser from navigating, and let's reset the
+      // signals so we start fresh on the next event.
+      event.preventDefault();
+      _triggerSignals.reset();
+      return;
+    }
+
+    if (_triggerSignals.isReadyToTrigger) {
+      _triggerLink();
+    }
   }
 
   /// Call this method to indicate that a hit test has been registered for the
   /// given [controller].
   ///
   /// The [onClick] callback is invoked when the anchor element receives a
   /// `click` from the browser.
-  static void registerHitTest(LinkViewController controller) {
-    _hitTestedViewId = controller.viewId;
-  }
+  static void onFollowLink(int viewId) {
+    _triggerSignals.registerFollowLink(viewId: viewId);
+
+    if (!_triggerSignals.isValid) {
+      // The view ID from follow link doesn't match the view ID from the event
+      // target. Let's prevent the browser from navigating, and let's reset the
+      // signals so we start fresh on the next event.
+      _triggerSignals.mouseEvent?.preventDefault();
+      _triggerSignals.reset();
+      return;
+    }
 
-  /// Removes all information about previously registered hit tests.
-  static void unregisterHitTest() {
-    _hitTestedViewId = null;
+    if (_triggerSignals.isReadyToTrigger) {
+      _triggerLink();
+    }
```

Comment:
The pattern of this `if` block looks the same everywhere. Can we pass the `_triggerLink` function to `LinkTriggerSignals` and have it call it at the appropriate time automatically? It already has the knowledge of when it is appropriate to call it. `LinkTriggerSignals` will no longer need `isReadyToTrigger` (except maybe for testing).

Reply:
There are cases where I also check for `isValid` first and do something else. But I think that also can be moved to the `LinkTriggerSignals` class. I like the idea so I'll take a stab at it.